### PR TITLE
*: resolve select fields properly for coalesced columns of natural join

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/25041

Problem Summary:

Incompatible error when grouping on join.

### What is changed and how it works?

What's Changed:

Specially mark the redundant columns incurred by natural join / `using` keyword in `FieldName`, and ignore them when resolving columns.

How it Works:

The problem was incurred by https://github.com/pingcap/tidb/pull/21922 when trying to fix https://github.com/pingcap/tidb/issues/6712. #21922 exposes the columns in `redundantSchema` to the schema of aggregation, which led to the ambiguity.

I have tried reverting the #21922 patch and finding the column in `LogicalJoin.redundantSchema` recursively when resolving the select fields, it can work in the plan building step, but would fail in the final `ResolveIndices`, because aggregation is a schema producer.

### Related changes

N/A

### Check List 

Tests 

- Unit test

Side effects

N/A

### Release note 

- Resolve select fields properly for coalesced columns of natural join

### CODE REVIEW

File: types/field_name.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ type FieldName struct {
 	// update stmt can write `writeable` column implicitly but cannot use non-public columns explicit.
 	// e.g. update t set a = 10 where b = 10; which `b` is in `writeOnly` state
 	NotExplicitUsable bool
+
+	Redundant bool
```
Comment (winoros) [inline types/field_name.go:38]: Maybe we could add some comments to illustrate the usage of it.

**Review by winoros — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**